### DISCUSSION TITLE
OpenGL acceleration with Docker

### DISCUSSION THREAD

**Discussion by conroy-cheers** [microsoft/wslg#395] [Q&A]
Has anyone managed to get an OpenGL accelerated Docker container running behind WSLG?
I can get GUI apps (e.g. Gazebo) running great with acceleration in a WSL2 Ubuntu 20.04 distro, just by upgrading Mesa. However, I haven't figured out how to do this in a container. My `docker-compose` config maps the following volumes and environment variables, and Gazebo runs in the container, but performance is poor, as only `llvmpipe` software rendering is being used.

      volumes:
        - /tmp/.X11-unix:/tmp/.X11-unix 
        - /mnt/wslg:/mnt/wslg
        - /usr/lib/wsl:/usr/lib/wsl
        - /dev/dxg:/dev/dxg
      environment:
        - DISPLAY
        - WAYLAND_DISPLAY
        - XDG_RUNTIME_DIR
        - PULSE_SERVER

I've tried this config with both Docker Desktop (which runs Docker in its own WSL2 distro), and Docker installed directly in my WSL2 Ubuntu 20.04 distro, but the results are identical either way.
What's the magic sauce? How does Mesa detect the d3d12 backend? According to [this blog post](https://devblogs.microsoft.com/directx/directx-heart-linux/#opengl-opencl-vulkan), the userspace just needs Mesa >=21.0, libd3d12/libdxcore, and /dev/dxg present. `libd3d12.so` and `libdxcore.so` are both mapped through /usr/lib/wsl, /dev/dxg is mapped, and Mesa 21.1.5 is installed in the container.

Any help would be greatly appreciated :)

**Comment by onomatopellan** [comment]
For GPU support inside the container from the command line you need to run docker with the argument `--gpus=all`.
For docker-compose you need to add the [gpu capabilities](https://docs.docker.com/compose/gpu-support/).

**Reply by conroy-cheers** [reply]
This works for CUDA, but doesn't seem to help for OpenGL

**Reply by onomatopellan** [reply]
I see. What's the output for `glxinfo -B` inside the container?

**Reply by dwbertol** [reply]
I'm facing the same problem trying to overcome it without success for a while. 
My `glxinfo -B` inside docker is

```
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Mesa/X.org (0xffffffff)
    Device: llvmpipe (LLVM 12.0.0, 256 bits) (0xffffffff)
    Version: 21.0.3
    Accelerated: no
    Video memory: 11898MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
OpenGL vendor string: Mesa/X.org
OpenGL renderer string: llvmpipe (LLVM 12.0.0, 256 bits)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 21.0.3
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 3.1 Mesa 21.0.3
OpenGL shading language version string: 1.40
OpenGL context flags: (none)

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 21.0.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```

**Comment by onomatopellan** [comment]
The team just posted a sample of a container with accelerated OpenGL. 

https://github.com/microsoft/wslg/tree/main/samples/container

After looking at that this worked for me:
```
docker run -it --rm -v /tmp/.X11-unix:/tmp/.X11-unix -v /mnt/wslg:/mnt/wslg -v /usr/lib/wsl:/usr/lib/wsl --device=/dev/dxg -e DISPLAY=$DISPLAY -e WAYLAND_DISPLAY=$WAYLAND_DISPLAY -e XDG_RUNTIME_DIR=$XDG_RUNTIME_DIR -e PULSE_SERVER=$PULSE_SERVER -e LD_LIBRARY_PATH=/usr/lib/wsl/lib --gpus all ubuntu/latest:opengl2
```

I guess the most important part was using the env variable LD_LIBRARY_PATH=/usr/lib/wsl/lib
